---
layout: default
title: Bismark file description
nav_order: 3
parent: Bismark manual
description: Description of the structure of files generated by Bismark
has_children: false
published: true
---


### Bismark BAM/SAM output (default)

By default, Bismark generates SAM output for all alignment modes. Please note that reported quality values are encoded in Sanger format (Phred 33 scale), even if the input was in Phred64.

1. `QNAME` (read name)
2. `FLAG` (this flag tries to take the strand a bisulfite read originated from into account (this is different from ordinary DNA alignment flags!))
3. `RNAME` (reference chromosome)
4. `POS` (start position)
5. `MAPQ` (calculated for Bowtie 2 and HISAT2)
6. `CIGAR`
7. `RNEXT`
8. `PNEXT`
9. `TLEN`
10. `SEQ`
11. `QUAL` (Phred33 scale)
12. `NM-tag` (edit distance to the reference)
13. `MD-tag` (base-by-base mismatches to the reference) (14) XM-tag (methylation call string)
14. `XR-tag` (read conversion state for the alignment) (16) XG-tag (genome conversion state for the alignment)

The mate read of paired-end alignments is written out as an additional separate line in the same format.

[Back to the tutorial](https://gabbo89.github.io/EEA2024/docs/3a1_WGBS_cleaning_and_alignment.html#bismark-bam)


### Methylation call

The methylation call string contains a dot `.` for every position in the BS-read not involving a cytosine, or contains one of the following letters for the three different cytosine methylation contexts (UPPER CASE = METHYLATED, lower case = unmethylated):

- `z` - C in CpG context - unmethylated
- `Z` - C in CpG context - methylated
- `x` - C in CHG context - unmethylated
- `X` - C in CHG context - methylated
- `h` - C in CHH context - unmethylated
- `H` - C in CHH context - methylated
- `u` - C in Unknown context (CN or CHN) - unmethylated
- `U` - C in Unknown context (CN or CHN) - methylated
- `.` - not a C or irrelevant position

[Back to the tutorial](https://gabbo89.github.io/EEA2024/docs/3a1_WGBS_cleaning_and_alignment.html#bismark-bam)
