---
layout: default
title: Lesson 2 - Oxford Nanopore Technologies alignment
nav_order: 2
parent: 3. Tutorial
description: A comprehensive guide to understanding epigenetics.
published: true
---
{: .important-title }
> Aim
>
> Perform an alignment of Oxford Nanopore reads to the reference genome. We will compare the results obtained with this sequencing technology with the ones obtained with Illumina sequencing technology.
> 


<br>
<details open markdown="block">
  <summary>
    <strong>Table of contents</strong>
  </summary>
  {: .text-delta }
- TOC
{:toc}
</details>
<br>


During this lesson we will focus on the alignemnt of ONT reads to the reference genome. ONT sequencing enable the direct detection of methylation levels, without the need of bisulfite conversion. The alignment of the reads is performed using `minimap2`. This tool is specifically designed for long-range sequencing data and is able to handle the unique characteristics of ONT data. 

---
Connect to the server and set the working directory
```bash
# Activate the conda environment
conda activate epigenomics

# Set the working directory
cd /data2/student_space/st24_16_folder/epigenomics/

# Create the required directories 
mkdir ont/sequences ont/reference

```
# 1. Prepare the reads for alignment
ONT sequencing data need to be in the appropriate format (usually `fastq` or `fasta` or `bam`). 
The output of nanopore sequencing may be available in different formats. 

We will use an input that is in bam format. We need to convert it to fastq [fastq format](../2a1_file_formats.html#fastq-format).

{: .highlight-title}
> Question
>
> How do we convert a bam to fastq?
>

<details>
    <summary>Show answer</summary>
We can take a look to the options of samtools.<br>
<br>

Open the short help of samtools 

```bash
samtools 
```

```
  -- File operations
     collate        shuffle and group alignments by name
     cat            concatenate BAMs
     consensus      produce a consensus Pileup/FASTA/FASTQ
     merge          merge sorted alignments
     mpileup        multi-way pileup
     sort           sort alignment file
     split          splits a file by read group
     quickcheck     quickly check if SAM/BAM/CRAM file appears intact
     fastq          converts a BAM to a FASTQ{: .text-red-100 }
     fasta          converts a BAM to a FASTA
     import         Converts FASTA or FASTQ files to SAM/BAM/CRAM
     reference      Generates a reference from aligned data
     reset          Reverts aligner changes in reads
```
```bash
samtools fastq
```

</details>


```bash
# Create the required directories 
mkdir ont/sequences ont/reference
```

The input file is located here 
`/data2/biotecnologie_molecolari_magris/epigenomics/ont/rkatsiteli.leaves.ont.bam`

MM:Z and ML:B optional bam tags are of paramount importance, because they hold the methylation information. They are generated by the ONT software that convert the signals to nucleotides.

For a detailed description refer to [SAM/BAM format specification](https://samtools.github.io/hts-specs/SAMtags.pdf), section **1.7 Base modifications**.

Each modified base prediction has a quality value associated with it, which should be interpreted as the likelihood of the modification being correct, given the assumption the original call is correct. 

#### `MM:Z`
{: .no_toc }

First character is the unmodified base reported by sequencing instrument for the top strand (`A`, `C`, `G`, `T`, `U` or `N`), followed by +/- indicating the strand the modification was observed (relative to the original strand of SEQ), and by one or more base modifications codes (for example *m* or *h*). 

![modifications ](image-4.png)

^1 for more details about ChEBI - Chemical Entities of Biological Interest (ChEBI) is a freely available dictionary of molecular entities focused on ‘small’ chemical compounds. [https://www.ebi.ac.uk/chebi/](https://www.ebi.ac.uk/chebi/)


Usually a `?` is following, and no information is available about the modification status of the skipped bases. Afterwards a comma separated list of how may bases in the read of the previously stated *base type* to skip (starting with 0). 

For example:

`C+m?,5,12,0` 

It tells us that there are 3 Cs called (as either modified or unmodified) on the top strand. The first 5 Cs are unknown, while the 6<sup>th</sup>, 19<sup>th</sup> and 20<sup>th</sup> are called Cs. 12 Cs between 6<sup>th</sup> and 19<sup>th</sup> are unknown. 

<!--
https://github.com/samtools/hts-specs/blob/a6a4504917a1b02197538f21e1b441c3f3892be4/SAMtags.tex#L517
-->

#### `ML:B`
{: .no_toc }

The tag lists the probability of each modification listed in the MM tag being correct, in the order they occour. Probability range between 0 and 1 and is remapped in equal sized portions to the discrete integers 0 to 255 inclusively. The probability range corresponding to integer value N is N/256 to (N+1)/256. 


```bash
# Create the required directories 
samtools fastq \
-@ ${threads} \
-T MM,ML \
/iga/vitis_sp/lanes/sequences/dna/vitis_vinifera/rkatsiteli/ONT_ultralong/ID2868_rkatsiteli/bam_pass/ID2868_Nanopore_Rkatsiteli_5mCG_5hmCG.bam

```




# 2. Align the reads to the reference genome 
We will use `minimap2` to align the reads to the reference genome. The reference genome need to be first indexed appropriately.





# 3. Extract methylation informations 
We will use `modkit` to extract the methylation informations from the aligned reads.  [Modkit short manual][Modkit short manual]{: .btn } and [Modkit on Github][Modkit_github]{: .btn }

## Create the bedMethyl tables 
A primary use of `modkit` is to create summary counts of modified and unmodified bases in an extended bedMethyl format. bedMethyl files tabulate the counts of base modifications from every sequencing read over each aligned reference genomic position. Only **primary alignments** are used in generating the table, it is recommended to mark duplicate alignments before running as multiple primary alignments can be double counted (but the behavior is logged).

In the simples usage modkit pileup creates a bedMethyl file from a pileup file. The pileup file is generated by the alignment tool, in this case `minimap2`. The pileup file is a text file that contains the aligned reads in a tabular format. The bedMethyl file is a binary file that contains the summary counts of modified and unmodified bases. The bedMethyl file can be used to visualize the methylation pattern using tools such as IGV. 


#### Usage
```bash
modkit pileup \
--cpg \
--filter-threshold 0.75 \
--only-tabs \
--combine-mods \
--combine-strands \
--ref ${ref_path} \
--log-filepath ${out_var}/dna_methylation/modkit/${sample}/${aligner}${sample}${hap}/modkit-pileup_${aligner}${sample}${hap}combinemods_combinestrands_ft075.CG.log \
-t ${threads} \
${out_path}/${sample}/${aligner}-${snp_caller}/snp_phasing/${aligner}${sample}.${hap}.bam \
${out_var}/dna_methylation/modkit/${sample}/${aligner}${sample}${hap}/${file_name}.${aligner}${sample}${hap}_combinemods_combinestrands_ft075.CG.bedMethyl 
```

#### the options 
- `--bowtie2 bowtie2` is used as the backend (DEFAULT).


{: .success-title }
> STDOUT
>
> Finished writing out cytosine report for covered chromosomes (processed 343 chromosomes/scaffolds in total)
>
> Now processing chromosomes that were not covered by any methylation calls in the coverage file...
>
> Writing cytosine report for not covered chromosome h1tg000079l
> Writing cytosine report for not covered chromosome h1tg000179l
> Finished writing out cytosine report (processed 345 chromosomes/scaffolds in total). coverage2cytosine processing complete.
>
> Finished generating genome-wide cytosine report!'

[Modkit short manual]: https://gabbo89.github.io/EEA2024-2025/docs/2a4_Modkit_manual.html
[Modkit_github]: https://github.com/nanoporetech/modkit

